File: 3/python/problem3.py
def divis(a: int, b: int) -> bool:
    return a % b == 0

def isprime(n: int) -> bool:
   return n == 2 or n == 3 or (n % 2 != 0 and n % 3 != 0 and n % 5 != 0 and\
         all(n % i != 0 and n % (i + 2) != 0 for i in range(5, int(n**0.5), 6)))

def wheel_factor(n: int) -> int:
    if isprime(n): return n

    inc = [4, 2, 4, 2, 4, 6, 2, 6]

    for prime in [2, 3, 5, 7, 11]:
        if divis(n, prime):
            return prime

    test = False

    k, i = 7, 0
    while not test and k * k <= n:
        test = divis(n, k)
        if test: return k
        k += inc[i]
        i = i + 1 if i < 7 else 0
    return n

n = 600851475143
print(n // wheel_factor(n))

File: 3/haskell/problem3.hs
import Data.Maybe
import Data.List (find)

divis :: Int -> Int -> Bool
divis n i = n `rem` i == 0

isPrime :: Int -> Bool
isPrime n
  | n == 1 = False
  | n `elem` [2, 3, 5] = True
  | otherwise = not $ any (divis n) (2 : 3 : [5,11..intSqrt n])
  where intSqrt n = ceiling $ sqrt $ fromIntegral n

wheelFactor :: Int -> Int
wheelFactor n
  | isPrime n = n
  | isJust found = fromJust found 
  where found = find (divis n) ([2, 3, 5, 7, 11] ++ testLS)
        testLS = 7 : 11 : zipWith (+) (tail $ cycle [4, 2, 4, 2, 4, 6, 2, 6]) (tail testLS)

main = do
    let n = 600851475143
    putStrLn $ show $ n `div` (wheelFactor n)
